trinity/bootstrap.py

```diff
@@ -0,0 +1,282 @@
+    "\n"
+    r"      ______     _       _ __       ",
+    r"     /_  __/____(_)___  (_) /___  __",
+    r"      / / / ___/ / __ \/ / __/ / / /",
+    r"     / / / /  / / / / / / /_/ /_/ / ",
+    r"    /_/ /_/  /_/_/ /_/_/\__/\__, /  ",
+    r"                           /____/   ",
+))
+
+TRINITY_AMBIGIOUS_FILESYSTEM_INFO = (
+    "Could not initialize data directory\n\n"
+    "   One of these conditions must be met:\n"
+    "   * HOME environment variable set\n"
+    "   * XDG_TRINITY_ROOT environment variable set\n"
+    "   * TRINITY_DATA_DIR environment variable set\n"
+    "   * --data-dir command line argument is passed\n"
+    "\n"
+    "   In case the data directory is outside of the trinity root directory\n"
+    "   Make sure all paths are pre-initialized as Trinity won't attempt\n"
+    "   to create directories outside of the trinity root directory\n"
+)
+
+
+BootFn = Callable[[
+    Namespace,
+    TrinityConfig,
+    Dict[str, Any],
+    PluginManager,
+    logging.handlers.QueueListener,
+    EventBus,
+    Endpoint,
+    logging.Logger
+], None]
+
+
+def main_entry(trinity_boot: BootFn) -> None:
+    event_bus = EventBus(ctx)
+    main_endpoint = event_bus.create_endpoint(MAIN_EVENTBUS_ENDPOINT)
+    main_endpoint.connect_no_wait()
+
+    plugin_manager = setup_plugins(
+        MainAndIsolatedProcessScope(event_bus, main_endpoint)
+    )
+    plugin_manager.amend_argparser_config(parser, subparser)
+    args = parser.parse_args()
+
+    if args.network_id not in PRECONFIGURED_NETWORKS:
+        raise NotImplementedError(
+            f"Unsupported network id: {args.network_id}.  Only the ropsten and mainnet "
+            "networks are supported."
+        )
+
+    has_ambigous_logging_config = (
+        args.log_levels is not None and
+        None in args.log_levels and
+        args.stderr_log_level is not None
+    )
+    if has_ambigous_logging_config:
+        parser.error(
+            "\n"
+            "Ambiguous logging configuration: The logging level for stderr was "
+            "configured with both `--stderr-log-level` and `--log-level`. "
+            "Please remove one of these flags",
+        )
+
+    if is_prerelease():
+        # this modifies the asyncio logger, but will be overridden by any custom settings below
+        enable_warnings_by_default()
+
+    stderr_logger, formatter, handler_stream = setup_trinity_stderr_logging(
+        args.stderr_log_level or (args.log_levels and args.log_levels.get(None))
+    )
+
+    if args.log_levels:
+        setup_log_levels(args.log_levels)
+
+    try:
+        trinity_config = TrinityConfig.from_parser_args(args)
+    except AmbigiousFileSystem:
+        parser.error(TRINITY_AMBIGIOUS_FILESYSTEM_INFO)
+
+    if not is_data_dir_initialized(trinity_config):
+        # TODO: this will only work as is for chains with known genesis
+        # parameters.  Need to flesh out how genesis parameters for custom
+        # chains are defined and passed around.
```

Comment:
Is this still accurate?

Reply:
I believe it is. I will soon dig deeper into this.